guides/keras_nlp/neural_machine_translation.py

```diff
@@ -0,0 +1,442 @@
+"""
+Title: English-to-Spanish translation with KerasNLP
+Author: [Abheesht Sharma](https://github.com/abheesht17/)
+Date created: 2022/05/26
+Last modified: 2022/05/26
+Description: Use KerasNLP to train a sequence-to-sequence Transformer model on the machine translation task.
+"""
+
+"""
+## Introduction
+
+KerasNLP provides building blocks for NLP (model layers, tokenizers, metrics, etc.) and
+makes it convenient to construct NLP pipelines.
+
+In this example, we'll use KerasNLP layers to build an encoder-decoder Transformer
+model, and train it on the English-to-Spanish machine translation task.
+
+You'll learn how to:
+
+- Tokenize text using the `WordPieceTokenizer` from KerasNLP.
+- Implement a sequence-to-sequence Transformer model using KerasNLP's `TransformerEncoder`
+, `TransformerDecoder` and `PositionalEmbedding` layers, and train it.
+- Use the trained model to generate translations of unseen input sentences!
+- Vectorize text using the Keras `TextVectorization` layer.
+
+Don't worry if you aren't familiar with KerasNLP. This tutorial shows exactly
+how simple and easy it is to use it! So, what are you waiting for? Dive right in!
+"""
+
+"""
+## Setup
+Before we start implementing the pipeline, let's import all the libraries we need.
+"""
+
+"""shell
+pip install -q keras-nlp
+pip install -q -U "tensorflow-text==2.8.*"
+"""
+
+import keras_nlp
+import numpy as np
+import pathlib
+import random
+import tensorflow as tf
+
+from tensorflow import keras
+from tensorflow_text.tools.wordpiece_vocab import bert_vocab_from_dataset as bert_vocab
```

Comment:
This is not a great UX... can we include it in KerasNLP?

Reply:
Hmmm, not sure. @mattdangerw, thoughts?

Reply:
I'm increasingly of the mind that we should wrap this tf text tool and the SentencePiece trainer (from SentencePiece pypi) in a universal trainer interface. It will really improve the UX here.

But I don't think we need to block this example on having it. IMO we should ship this using the tf text tool, and when we have a trainer class in KerasNLP replace the invocation here.

Reply:
>>> from tensorflow_text.tools.wordpiece_vocab import bert_vocab_from_dataset as bert_vocab

>> This is is not great UX 

> we should wrap this tf text tool and the SentencePiece trainer (from SentencePiece pypi) in a universal trainer interface.
>
> But I don't think we need to block this example on having it.

+1. 

I rigged that up (for [this tutorial](https://www.tensorflow.org/text/guide/subwords_tokenizer)) back when there was _no_ public interface for learning a wordpiece vocab. I tried, but failed, to convince TFText team that this should probably be put in the `.adapt` method of a `WordpieceTokenizer` class. Their reasons, IIRC, were mainly "you shouldn't use `.adapt` for any real problem. You'll want to use an apache-beam pipeline".

I find that one of the good arguments for keeping the vocabulary learner close to the tokenizer is just to be sure you're using the same normalization and splitting in both.

Reply:
What's the current status here?

Reply:
I think the plan is to add a wrapper for this in KerasNLP (not sure), but we plan to roll with the current code for now, and update it later once it is a part of KerasNLP. @mattdangerw, please correct me if I am wrong.

Reply:
Yep! I think that's the plan right now!